File: 6_kyu/Split_Strings.go
package kata

// Splits str into strings of 2 characters.
// Adds a '_' if the number of chars in the str is odd.
func Solution(str string) []string {
	// Allow multibyte-character string: work with runes.
	r := []rune(str)
	len_r := len(r)
	// Round up the number of strings
	n := (len_r + 1) / 2
	result := make([]string, 0, n)
	for i := 0; i < len_r; i += 2 {
		result = append(result, string(r[i:i+2]))
	}
	if len_r&1 == 1 {
		// odd str length: need to fix the last string.
		result[n-1] = string([]rune{r[len_r-1], '_'})
	}
	return result
}

File: 6_kyu/Split_Strings.rs
// Splits s into strings of 2 characters.
// Appends '_' if needed.
fn solution(s: &str) -> Vec<String> {
    // Split s into characters (not into bytes).
    let a: Vec<char> = s.chars().collect();
    // Form string of 2 characters
    let mut accu: Vec<String> = Vec::with_capacity((a.len() + 1) / 2);
    for c in a.chunks_exact(2) {
        accu.push(format!("{}{}", c[0], c[1]))
    }
    // optional last character with a "_"
    if a.len() & 1 == 1 {
        accu.push(format!("{}_", a[a.len() - 1]))
    }
    accu
}
